go/fixchain/logger.go

```diff
@@ -0,0 +1,250 @@
+package fixchain
+
+import (
+	"encoding/json"
+	"io/ioutil"
+	"log"
+	"net/http"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/google/certificate-transparency/go/x509"
+)
+
+// Limiter is an interface to allow different rate limiters to be used with the
+// Logger.
+type Limiter interface {
+	Wait()
+}
+
+// Logger contains methods to asynchronously log certificate chains to a
+// Certificate Transparency log and properties to store information about each
+// attempt that is made to post a certificate chain to said log.
+type Logger struct {
+	url    string
+	client *http.Client
+	roots  *x509.CertPool
+	toPost chan *toPost
+	errors chan<- *FixError
+
+	active uint32
+	// Counters may not be entirely accurate due to non-atomicity.
```

Comment:
It did break when I ran my tests with the -race flag.  I have now made all of the counters be incremented atomically.